core/base_service.go

```diff
@@ -150,60 +151,87 @@ func (service *BaseService) SetUserAgent(userAgentString string) {
 }
 
 // Request performs the HTTP request
-func (service *BaseService) Request(req *http.Request, result interface{}) (*DetailedResponse, error) {
-	// Add default headers
+func (service *BaseService) Request(req *http.Request, result interface{}) (detailedResponse *DetailedResponse, err error) {
+	// Add default headers.
 	if service.DefaultHeaders != nil {
 		for k, v := range service.DefaultHeaders {
 			req.Header.Add(k, strings.Join(v, ""))
 		}
 	}
 
-	// Check if user agent is present.
+	// Add the default User-Agent header if no already present.
 	userAgent := req.Header.Get(USER_AGENT)
 	if userAgent == "" {
 		req.Header.Add(USER_AGENT, service.UserAgent)
 	}
 
 	// Add authentication to the outbound request.
 	if service.Options.Authenticator == nil {
-		return nil, fmt.Errorf(ERRORMSG_NO_AUTHENTICATOR)
+		err = fmt.Errorf(ERRORMSG_NO_AUTHENTICATOR)
+		return
 	}
-	err := service.Options.Authenticator.Authenticate(req)
+
+	err = service.Options.Authenticator.Authenticate(req)
 	if err != nil {
-		return nil, err
+		return
 	}
 
 	// Perform the request.
-	resp, err := service.Client.Do(req)
+	httpResponse, err := service.Client.Do(req)
 	if err != nil {
-		return nil, err
+		return
 	}
 
-	response := new(DetailedResponse)
-	response.Headers = resp.Header
-	response.StatusCode = resp.StatusCode
-	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
-		if resp != nil {
-			response.Result = resp
-			message := getErrorMessage(resp)
-			return response, fmt.Errorf(message)
-		}
+	detailedResponse = &DetailedResponse{
+		StatusCode: httpResponse.StatusCode,
+		Headers:    httpResponse.Header,
 	}
 
-	contentType := resp.Header.Get(CONTENT_TYPE)
-	if contentType != "" {
-		if IsJSONMimeType(contentType) && result != nil {
-			json.NewDecoder(resp.Body).Decode(&result)
-			response.Result = result
-			defer resp.Body.Close()
+	// Save off the Content-Type header value.
+	contentType := httpResponse.Header.Get(CONTENT_TYPE)
+
+	// Be sure to close the response body's Reader before we return.
+	defer httpResponse.Body.Close()
+
+	// If the operation was unsuccessful, then set up the DetailedResponse
+	// and error objects appropriately.
+	if httpResponse.StatusCode < 200 || httpResponse.StatusCode >= 300 {
+		var message string
+		if IsJSONMimeType(contentType) {
+			// Response body contains JSON.
+			// Return the error response body as a map, along with an
+			// error object containing our best guess at an error message.
+			responseMap, _ := decodeAsMap(httpResponse.Body)
+			detailedResponse.Result = responseMap
+			message = getErrorMessage(responseMap)
+		} else {
+			// Response body contains something other than JSON.
+			// Return the response body as a byte array in the RawResult field
+			// and also return the entire response body as an error message.
+			detailedResponse.RawResult, _ = ioutil.ReadAll(httpResponse.Body)
+			message = string(detailedResponse.RawResult)
```

Comment:
In Swift, we create a message from the `StatusCode`. Fortunately, there is a built-in function for that in Swift and we just use it. If there was something like that in Go, I'd recommend using it. If not, perhaps we could do something simple, like just looking at the "class" of the error (4xx vs 5xx) and setting a message from that.

Reply:
done